### PR TITLE
MINOR: Improvements in group metadata cleanup and test coverage

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

**Comment by hachikuji** [comment]
@ijuma @guozhangwang Some small improvements from the patch for KAFKA-4399 with some additional test coverage. Not required, but if it seems reasonable, maybe we can put in 0.10.1.

File: core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala [hunk 1]
```diff
@@ -561,79 +560,62 @@ class GroupMetadataManager(val brokerId: Int,
...
+
+            // We avoid writing the tombstone when the generationId is 0, since this group is only using
+            // Kafka for offset storage.
+            if (groupIsDead && groupMetadataCache.remove(groupId, group) && generation > 0) {
+              // Append the tombstone messages to the partition. It is okay if the replicas don't receive these (say,
+              // if we crash or leaders move) since the new leaders will still expire the consumers with heartbeat and
+              // retry removing this group.
+              tombstones += new Message(bytes = null, key = GroupMetadataManager.groupMetadataKey(group.groupId),
+                timestamp = timestamp, magicValue = magicValue)
+              trace(s"Marked group $groupId in $appendPartition for deletion")
```
Comment (ijuma) [inline core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala]: Nit: is it worth mentioning that it has been removed from the cache (we used to do that later, but we now do it here)?

File: core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala [hunk 2]
```diff
@@ -561,79 +560,62 @@ class GroupMetadataManager(val brokerId: Int,
...
+            if (tombstones.nonEmpty) {
+              try {
+                // do not need to require acks since even if the tombstone is lost,
+                // it will be appended again in the next purge cycle
+                partition.appendMessagesToLeader(new ByteBufferMessageSet(config.offsetsTopicCompressionCodec, tombstones: _*))
+                offsetsRemoved += expiredOffsets.size
+                trace(s"Successfully appended tombstones to $appendPartition for expired offsets and/or metadata for group $groupId")
+              } catch {
+                case t: Throwable =>
+                  error(s"Failed to append tombstones to $appendPartition for expired offsets and/or metadata for group $groupId.", t)
```
Comment (ijuma) [inline core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala]: We used to report the number of tombstones, is there a reason why you removed that?

**Review by ijuma — APPROVED** [review]
A couple of logging nits, LGTM otherwise. @guozhangwang, do you want to take a look as well since you are more familiar with this code?

File: core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala (hunk 1 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala]: Ack. It seems useful to know, but maybe at DEBUG or TRACE level?

File: core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala (hunk 2 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala]: Ack. Probably a casualty of moving things around.

**Comment by ijuma** [comment]
Thanks @hachikuji, LGTM.

**Comment by guozhangwang** [comment]
Discussed offline with @hachikuji. LGTM. Merged to trunk and 0.10.1.

**Pull request closed without merging**